astarte/device/database.py

```diff
@@ -0,0 +1,235 @@
+
+    @abstractmethod
+    def store_prop(self, interface: str, major: int, path: str, value: object) -> None:
+        """
+        Store a property value in the database. It will overwrite the previous value where present.
+
+        Parameters
+        ----------
+        interface : str
+            The interface name.
+        major : int
+            The major version for the interface.
+        path : str
+            The path to the property endpoint.
+        value : object
+            The new value for the property.
+        """
+
+    @abstractmethod
+    def load_prop(self, interface: str, major: int, path: str) -> object | None:
+        """
+        Load a property from the database. If a property is found but the major version does not
+        match, the property in the database will be deleted and None will be returned.
+
+        Parameters
+        ----------
+        interface : str
+            The interface name.
+        major : int
+            The major version for the interface.
+        path : str
+            The path to the property endpoint.
+
+        Returns
+        -------
+        object | None
+            The property value if the property is present and the provided interface major
+            version matches the interface version stored in the database. None otherwise.
+        """
+
+    @abstractmethod
+    def delete_prop(self, interface: str, path: str) -> None:
+        """
+        Delete a property from the database.
+
+        Parameters
+        ----------
+        interface : str
+            The interface name.
+        path : str
+            The path to the property endpoint.
+        """
+
+    @abstractmethod
+    def clear(self) -> None:
+        """
+        Fully clear the database of all the properties.
+        """
+
+    @abstractmethod
+    def load_all_props(self) -> list[tuple[str, int, str, object]]:
+        """
+        Load all the properties stored in the database.
+
+        Returns
+        -------
+        list[tuple[str, int, str, object]]
+            A list containing all the propeties stored in the database.
+            Each element of the list is a tuple in the format:
+            (interface, interface major version, path, value)
+        """
+
+
+class AstarteDatabaseSQLite(AstarteDatabase):
+    """
+    An implementation for the abstract AstarteDatabase class. This implementation uses the standard
+    SQLite library of python to implement property persistency.
+    """
+
+    def __init__(self, database_path: Path) -> None:
+        """
+        Parameters
+        ----------
+        database_path : Path
+            The path to the file to use to instantiate the database.
+        """
+        self.__database_path = database_path
+        cursor = sqlite3.connect(self.__database_path).cursor()
+        cursor.execute(
+            "CREATE TABLE IF NOT EXISTS properties "
+            "(interface TEXT, major INTEGER NOT NULL, path TEXT, value BLOB NOT NULL, "
```

Comment:
Adding `major` as part of the primary key could cause the duplication of properties for different versions.
It might be better to keep the primary key without the major.